flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/partitioner/RebalancePartitioner.java

```diff
@@ -33,29 +33,18 @@
 public class RebalancePartitioner<T> extends StreamPartitioner<T> {
 	private static final long serialVersionUID = 1L;
 
-	private final int[] returnArray = {Integer.MAX_VALUE - 1};
+	private final int[] returnArray = { -1 };
```

Comment:
Yes, I also considered to use a boolean `isInitialized` before as you said. In order not to bring another extra variable, I relies on `-1` directly. I agree another boolean value seems more cleaner than current way.

You are right that the current `ChannelSelector` is short of mechanism to initialization, and it may be necessary to add this logic in the interface. I will double check how many implementations are needing this logic or just `RebalancePartitioner`, especially for the batch implementations in `OutputEmitter`. Then I would consider whether to cover this refactor in this PR or in a separate PR next. :)

Reply:
Thanks for looking into it @zhijiangW :)

I think that adding an extra boolean flag is not necessary. It would be nice to add a `checkState(someConditionThatEnsuresSelectorWasInitialized)`, but not strictly speaking necessary as long as it's tested.

I don't think that we need `numChannels` in the `selectChannels()` method. I might be wrong, but I guess it was just a hack to workaround the lack of initialization. After all doesn't the `numChannels` always come from `RecordWriter`'s final field
```
	private final int numChannels;
```
?

Reply:
Yes, the `numChannels` always coms form the RecordWriter's final field, so we can call `ChannelSelector#setup(int numChannels)` then.

I think we are on the same page for the following:
1. It is no need to retain `numChannels` in the `selectChannels` method.
2. No need to introduce another boolean to check initialized state in `selectChannels` process.

They are just what I thought before, and glad to see you also think so. :)